internal/http/services/owncloud/ocdav/tpc.go

```diff
@@ -0,0 +1,418 @@
+}
+
+func (wc *WriteCounter) Write(p []byte) (int, error) {
+
+	n := len(p)
+	wc.Total += uint64(n)
+	NowTime := time.Now()
+
+	diff := NowTime.Sub(wc.PrevTime).Seconds()
+	if diff >= PerfMarkerResponseTime {
+		wc.SendPerfMarker(wc.Total)
+		wc.PrevTime = NowTime
+	}
+	return n, nil
+}
+
+//
+// An example of an HTTP TPC Pull
+//
+// +-----------------+        GET          +----------------+
+// |  Src server     |  <----------------  |  Dest server   |
+// |  (Remote)       |  ---------------->  |  (Reva)        |
+// +-----------------+       Data          +----------------+
+// 												^
+// 												|
+// 												| COPY
+// 												|
+// 										   +----------+
+// 										   |  Client  |
+// 										   +----------+
+
+// handleTPCPull performs a GET request on the remote site and upload it
+// the requested reva endpoint.
+func (s *svc) handleTPCPull(ctx context.Context, w http.ResponseWriter, r *http.Request, ns string) {
+	src := r.Header.Get("Source")
+	dst := path.Join(ns, r.URL.Path)
+	overwrite := r.Header.Get("Overwrite")
+	depth := r.Header.Get("Depth")
+	if depth == "" {
+		depth = "infinity"
+	}
+
+	sublog := appctx.GetLogger(ctx).With().Str("src", src).Str("dst", dst).Logger()
+	sublog.Debug().Str("overwrite", overwrite).Str("depth", depth).Msg("TPC Pull")
+
+	overwrite = strings.ToUpper(overwrite)
+	if overwrite == "" {
+		overwrite = "T"
+	}
+
+	if overwrite != "T" && overwrite != "F" {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	if depth != "infinity" && depth != "0" {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// get Gateway client
+	client, err := s.getClient()
+	if err != nil {
+		sublog.Error().Err(err).Msg("error getting grpc client")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// check if destination exists
+	ref := &provider.Reference{Path: dst}
+	dstStatReq := &provider.StatRequest{Ref: ref}
+	dstStatRes, err := client.Stat(ctx, dstStatReq)
+	if err != nil {
+		sublog.Error().Err(err).Msg("error sending grpc stat request")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	if dstStatRes.Status.Code != rpc.Code_CODE_OK && dstStatRes.Status.Code != rpc.Code_CODE_NOT_FOUND {
+		HandleErrorStatus(&sublog, w, dstStatRes.Status)
+		return
+	}
+
+	if overwrite == "F" {
+		sublog.Warn().Str("overwrite", overwrite).Msg("Destination already exists")
+		w.WriteHeader(http.StatusPreconditionFailed) // 412, see https://tools.ietf.org/html/rfc4918#section-9.8.5
+		return
+	}
+
+	w.WriteHeader(http.StatusAccepted)
+
+	err = s.performHTTPPull(ctx, client, r, w, depth == "infinity", ns)
+	if err != nil {
+		sublog.Error().Err(err).Str("depth", depth).Msg("error descending directory")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	fmt.Fprintf(w, "success: Created")
+}
+
+func (s *svc) performHTTPPull(ctx context.Context, client gateway.GatewayAPIClient, r *http.Request, w http.ResponseWriter, recurse bool, ns string) error {
```

Comment:
Cases where depth is infinity are not handled and will result in errors

Reply:
Good catch. I learned that for HTTP TPC depth is not relevant (!) and folders are not supported, therefore I'll modify the code to return `HttpBadRequest` when the target is a folder and will actually remove any reference to `depth` (including the not-implemented `recurse` parameter).